src/databricks/labs/ucx/source_code/jobs.py

```diff
@@ -441,3 +458,76 @@ def _process_dependency(
+
+    def _process_dependency(
+        self, dependency: Dependency, path_lookup: PathLookup, inherited_tree: Tree | None
+    ) -> Iterable[DirectFsAccess]:
+        language = file_language(dependency.path)
+        if not language:
+            logger.warning(f"Unknown language for {dependency.path}")
+            return
+        cell_language = CellLanguage.of_language(language)
+        source = dependency.path.read_text(guess_encoding(dependency.path))
+        if is_a_notebook(dependency.path):
+            yield from self._collect_from_notebook(source, cell_language, dependency.path, inherited_tree)
+        elif dependency.path.is_file():
+            yield from self._collect_from_source(source, cell_language, dependency.path, inherited_tree)
+
+    def _collect_from_notebook(
+        self, source: str, language: CellLanguage, path: Path, inherited_tree: Tree | None
+    ) -> Iterable[DirectFsAccess]:
+        notebook = Notebook.parse(path, source, language.language)
+        for cell in notebook.cells:
+            src_timestamp = int(path.stat().st_mtime)
+            src_id = str(path)
+            src_lineage = self.lineage_str
+            for dfsa in self._collect_from_source(cell.original_code, cell.language, path, inherited_tree):
+                yield dfsa.replace(
+                    source_type="NOTEBOOK", source_id=src_id, source_lineage=src_lineage, source_timestamp=src_timestamp
+                )
+            if cell.language is CellLanguage.PYTHON:
+                if inherited_tree is None:
+                    inherited_tree = Tree.new_module()
+                tree = Tree.normalize_and_parse(cell.original_code)
+                inherited_tree.append_tree(tree)
+
+    def _collect_from_source(
+        self, source: str, language: CellLanguage, path: Path, inherited_tree: Tree | None
+    ) -> Iterable[DirectFsAccess]:
+        iterable: Iterable[DirectFsAccess] | None = None
+        if language is CellLanguage.SQL:
+            iterable = self._collect_from_sql(source)
+        if language is CellLanguage.PYTHON:
+            iterable = self._collect_from_python(source, inherited_tree)
+        if iterable is None:
+            logger.warning(f"Language {language.name} not supported yet!")
+            return
+        src_id = str(path)
+        src_lineage = self.lineage_str
+        src_timestamp = int(path.stat().st_mtime)
+        for dfsa in iterable:
+            yield dfsa.replace(
+                source_type="FILE", source_id=src_id, source_lineage=src_lineage, source_timestamp=src_timestamp
+            )
+
+    def _collect_from_python(self, source: str, inherited_tree: Tree | None) -> Iterable[DirectFsAccess]:
+        linter = DirectFsAccessPyLinter(self._session_state, prevent_spark_duplicates=False)
+        for dfsa_node in linter.collect_dfsas(source, inherited_tree):
+            yield dfsa_node.dfsa
+
+    def _collect_from_sql(self, source: str) -> Iterable[DirectFsAccess]:
```

Comment:
do we analyse the `spark.sql("SELECT * FROM csv.`/mnt/foo/bar`")`?

Reply:
yes that's done as part of pyspark linting

Reply:
Actually it's not. Fixed in #2558